optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/tri/DroolsJoinTriConstraintStream.java

```diff
@@ -0,0 +1,112 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.stream.drools.tri;
+
+import java.util.UUID;
+
+import org.drools.model.Declaration;
+import org.drools.model.PatternDSL;
+import org.optaplanner.core.api.score.stream.tri.TriJoiner;
+import org.optaplanner.core.impl.score.stream.common.JoinerType;
+import org.optaplanner.core.impl.score.stream.drools.DroolsConstraintFactory;
+import org.optaplanner.core.impl.score.stream.drools.bi.DroolsAbstractBiConstraintStream;
+import org.optaplanner.core.impl.score.stream.drools.uni.DroolsAbstractUniConstraintStream;
+import org.optaplanner.core.impl.score.stream.tri.AbstractTriJoiner;
+
+public final class DroolsJoinTriConstraintStream<Solution_, A, B, C>
+        extends DroolsAbstractTriConstraintStream<Solution_, A, B, C> {
+
+    private final DroolsAbstractBiConstraintStream<Solution_, A, B> leftParentStream;
+    private final DroolsAbstractUniConstraintStream<Solution_, C> rightParentStream;
+    private final AbstractTriJoiner<A, B, C> triJoiner;
+    private final PatternDSL.PatternDef<C> cPattern;
+
+    public DroolsJoinTriConstraintStream(DroolsConstraintFactory<Solution_> constraintFactory,
+            DroolsAbstractBiConstraintStream<Solution_, A, B> parent,
+            DroolsAbstractUniConstraintStream<Solution_, C> otherStream, TriJoiner<A, B, C> triJoiner) {
+        super(constraintFactory, null);
+        this.leftParentStream = parent;
+        this.rightParentStream = otherStream;
+        this.triJoiner = (AbstractTriJoiner<A, B, C>) triJoiner;
+        this.cPattern = otherStream.getPattern().expr("triJoin-" + UUID.randomUUID(), getAVariableDeclaration(),
+                getBVariableDeclaration(), (c, a, b) -> matches(a, b, c));
+    }
+
+    // ************************************************************************
+    // Getters/setters
+    // ************************************************************************
+
+    @Override
+    public DroolsAbstractBiConstraintStream<Solution_, A, B> getLeftParentStream() {
+        return leftParentStream;
+    }
+
+    @Override
+    public DroolsAbstractUniConstraintStream<Solution_, C> getRightParentStream() {
+        return rightParentStream;
+    }
+
+    @Override
+    public Declaration<A> getAVariableDeclaration() {
+        return leftParentStream.getLeftVariableDeclaration();
+    }
+
+    @Override
+    public PatternDSL.PatternDef<A> getAPattern() {
+        return leftParentStream.getLeftPattern();
```

Comment:
Doesn't DroolsAbstractBiConstraintStream have getAPattern()?

If Tri's have getAPattern(), getBPattern() and getCPattern()
then Bi's must have getAPattern(), getBPattern() - but not getLeftPattern() and getRightPattern()
and Uni's getAPattern()
and one day Quads will have getAPattern(), getBPattern() getCPattern() and getDPattern()

Reply:
Notice that on Tri's there is no real "left" and "right": it is wrong to presume that the left is A+B and right is C, because the left might be A and the right might be C.
As for Bi's: they might be created from something else than a join (think a Tri.groupBy(key1, key2)), so the calling A left and calling B right starts breaking down.